### PR TITLE
feat(fsdp): add training option for fsdp

### PR DESCRIPTION
**Pull request by zaglc**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This PR adds interface for FSDP, which can be a subsitution of ZeRO1 and is compatible with TENSOR parallel. Running FSDP with PIPELINE parallel is not supported in this PR.

## Modification

+ add another group initializer `Initializer_Zero3_dp` in case that ZeRO3 parallel size is less than DP size and all-reduce is needed;
+ add an simplified optimizer `FSDPadaptOptimizer` because FSDP module will automatically accompolish most of the communication;
+ add interface for FSDP model storing in order to be compatible with current framework;
+ add interface of FSDP and warp the original model;

## Use cases (Optional)

If you want to train the model with FSDP, set the `use_fsdp` flag in parallel dict in 7B_sft.py like this:
```
parallel = dict(
    zero1=-1,
    pipeline=dict(size=1, interleaved_overlap=True),
    tensor=2,
    sequence_parallel=False,
    use_fsdp = True,
)
```

## Checklist

+ The PR requires more test for some reasons such as loss value alignment with original parallelized training method
+ May commit other changes basing on this PR in the future

### CODE REVIEW

**Comment by zaglc** [comment]
## Supplementary explanation (1)
Currently,  for FSDP combined with tensor parallel:
+ In Implementation of model part of `InternLM`, such as `internlm/model/linear.py, Line 200`, an extra attribute `IS_TENSOR_PARALLEL` is directly set in some model parameters,
  ```
          if gpc.get_world_size(ParallelMode.TENSOR) > 1:
              for name in ["w1", "w2", "w3"]:
                  for param in getattr(self, name).parameters():
                      setattr(param, IS_TENSOR_PARALLEL, True)
  ```
+ FSDP use a large flatten tensor to manage all the parameters within a FSDP unit, and the `detach` method is used, so attribute `IS_TENSOR_PARALLEL` will be lost and lead to wrong result when computing norm,
  ```
          with torch.no_grad():
              flat_params = [
                  p.detach().reshape(-1) if isinstance(p, nn.Parameter) else p.reshape(-1)
                  for p in params
              ]
              flat_param_data = torch.cat(flat_params, dim=0)
          flat_param = FlatParameter(flat_param_data, requires_grad=requires_grad)
          return flat_param
  ```
+ To reset this attribute, another attribute of FSDP module called `_shard_param_offsets` can be used, but it may be depreciated in the future.
